torchtext/data/functional.py

```diff
@@ -40,6 +41,29 @@ def generate_sp_model(filename, vocab_size=20000,
     torch.ops.torchtext.generate_sp_model(filename, vocab_size, model_type, model_prefix)
 
 
+def generate_sp_model_from_iterator(lines, vocab_size=20000,
+                                    model_type="unigram",
+                                    model_prefix='m_user'):
+    r"""Train a SentencePiece tokenizer.
+
+    Args:
+        lines: an iterator containing data for training SentencePiece model.
+        vocab_size: the size of vocabulary (Default: 20,000).
+        model_type: the type of SentencePiece model, including unigram,
+            bpe, char, word.
+        model_prefix: the prefix of the files saving model and vocab.
+
+    Outputs:
+        The model and vocab are saved in two separate files with
+            model_prefix.
+
+    Examples:
+        >>> from torchtext.data.functional import generate_sp_model_from_iterator
+        >>> generate_sp_model_from_iterator(lines, vocab_size=23456, model_prefix='spm_user')
+    """
+    return _generate_sp_model_from_iterator(lines, vocab_size, model_type, model_prefix)
```

Comment:
I think this can be combined with the existing `generate_sp_model`.

```python
def generate_sp_model(...):
    if not torch.jit.is_scripting():
        if not <the first arg is path-like object>
            return torchtext._torchtext._generate_sp_model_from_iterator(...)
    return torch.ops.torchtext.generate_sp_model(...)
```

Reply:
I like this approach, but one concern I have is that the first arg of `generate_sp_model` is `filename` -- it would be misleading to users to have an iterable passed with name "filename" and changing it is a breaking change (albeit a silly one since it's a positional arg). WDYT?